addons/mass_mailing/models/mass_mailing_stats.py

```diff
@@ -15,118 +15,100 @@ class MailMailStats(osv.Model):
     _rec_name = 'message_id'
     _order = 'message_id'
 
-    def _compute_state(self, cr, uid, ids, field_names, arg, context=None):
-        res = dict((i, {'state': 'outgoing', 'state_update': fields.datetime.now()}) for i in ids)
-
-        for stat in self.browse(cr, uid, ids, context=context):
+    mail_mail_id = fields.Many2one('mail.mail', string='Mail', index=True)
+    mail_mail_id_int = fields.Integer(
+        string='Mail ID (tech)',
+        help='ID of the related mail_mail. This field is an integer field because '
+             'the related mail_mail can be deleted separately from its statistics. '
+             'However the ID is needed for several action and controllers.'
+    )
+    message_id = fields.Char(string='Message-ID')
+    model = fields.Char(string='Document model')
+    res_id = fields.Integer(string='Document ID')
+    # campaign / wave data
+    mass_mailing_id = fields.Many2one('mail.mass_mailing', string='Mass Mailing')
+    mass_mailing_campaign_id = fields.Many2one(
+        related='mass_mailing_id.mass_mailing_campaign_id',
+        string='Mass Mailing Campaign',
+        store=True, readonly=True)
+    # Bounce and tracking
+    scheduled = fields.Datetime(help='Date when the email has been created', default=fields.Datetime.now)
+    sent = fields.Datetime(help='Date when the email has been sent')
+    exception = fields.Datetime(help='Date of technical error leading to the email not being sent')
+    opened = fields.Datetime(help='Date when the email has been opened the first time')
+    replied = fields.Datetime(help='Date when this email has been replied for the first time.')
+    bounced = fields.Datetime(help='Date when this email has bounced.')
+    links_click_ids = fields.One2many('link.tracker.click', 'mail_stat_id', string='Links click')
+    state = fields.Selection(compute="_compute_state",
+                             selection=[('outgoing', 'Outgoing'),
+                                        ('exception', 'Exception'),
+                                        ('sent', 'Sent'),
+                                        ('opened', 'Opened'),
+                                        ('replied', 'Replied'),
+                                        ('bounced', 'Bounced')], store=True)
+    state_update = fields.Datetime(compute="_compute_state", string='State Update',
+                                    help='Last state update of the mail',
+                                    store=True)
+    recipient = fields.Char(compute="_compute_recipient")
+
+    @api.depends('sent', 'opened', 'replied', 'bounced', 'exception')
+    def _compute_state(self):
+        self.update({'state': 'outgoing', 'state_update': fields.Datetime.now()})
```

Comment:
This will not work !

Before you were setting the state by default at 'outgoing' before commiting this change on the records (before `return res`)

Now you are setting all the states at 'outgoing', so during the `for stat in self`, you will never enter into an if statement. This is a correct way to do it:
``` python
    @api.depends('sent', 'opened', 'replied', 'bounced', 'exception')
    def _compute_state(self):
        self.update({'state_update': fields.Datetime.now()})
        for stat in self:
            if stat.exception:
                stat.state = 'exception'
            elif stat.sent:
                stat.state = 'sent'
            elif stat.opened:
                stat.state = 'opened'
            elif stat.replied:
                stat.state = 'replied'
            elif stat.bounced:
                stat.state = 'bounced'
            else:
                stat.state = 'outgoing'
```